Họ tên: Nguyễn Anh Khôi

MSSV: 22127208

Link dẫn đến file Google Colab: https://colab.research.google.com/drive/1KtrQERVSueYIhg7EbaCAO9T9J52EQYGW?usp=sharing

# HW1: Giới thiệu CUDA

Với các GPU tương đối mới thì để biên dịch chỉ cần dùng câu lệnh: \
`nvcc tên-file.cu -o tên-file-chạy`

Nhưng trên Colab mình thường lấy được GPU khá cũ là Tesla K80 với compute capability (phiên bản phần cứng) là 3.7; để biên dịch đúng với GPU khá cũ này thì bạn cần dùng câu lệnh: \
`nvcc -arch=sm_37 tên-file.cu -o tên-file-chạy` \
Trong đó, 37 chính là compute capability của GPU Tesla K80.

Để phòng trường hợp khi làm bài bạn lấy được GPU có compute capability x.x nhưng khi chấm bài Thầy lại lấy được GPU có compute capability khác x.x, dưới đây mình sẽ có đoạn code Python để tự động lấy 2 con số ứng với compute capability của GPU và lưu vào 2 biến `major` và `minor`:


In [2]:
from numba import cuda
major, minor = cuda.get_current_device().compute_capability
print(f'GPU compute capability: {major}.{minor}')

GPU compute capability: 7.5


Một khi đã chạy đoạn code Python ở trên, để biên dịch thì bạn sẽ dùng câu lệnh: \
`nvcc -arch=sm_{major}{minor} tên-file.cu -o tên-file-chạy`

Dưới đây, khi làm bài thì bạn có thể tùy ý thêm/xóa cell. Đừng xóa mấy cell có chữ của Thầy là được.

## Câu 1

In [3]:
!nvcc -arch=sm_{major}{minor} HW1_P1.cu -o HW1_P1.out

In [4]:
!./HW1_P1.out ./in.pnm output_16_16 16 16

Image size (width x height): 512 x 512

Processing time (use host): 3.702784 ms

GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use device): 1.824608 ms

Error between device result and host result: 0.000000


In [5]:
!./HW1_P1.out ./in.pnm output_32_32 32 32

Image size (width x height): 512 x 512

Processing time (use host): 3.715072 ms

GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use device): 2.726080 ms

Error between device result and host result: 0.000000


In [6]:
!./HW1_P1.out ./in.pnm output_64_64 64 64

Image size (width x height): 512 x 512

Processing time (use host): 3.688416 ms

GPU name: Tesla T4
GPU compute capability: 7.5
Error: HW1_P1.cu:184, code: 9, reason: invalid configuration argument


Chạy với `(blockSize.x, blockSize.y) = (64,64)` thì ta sẽ nhận được lỗi `Error: HW1_P1.cu:184, code: 9, reason: invalid configuration argument`. Dòng số 184 trong file `HW1_P1.cu` là dòng `CHECK(cudaGetLastError());` ngay sau lệnh gọi kernel. Lỗi này là do số lượng thread mà ta yêu cầu cấp phát cho một block (64x64=4096 threads/block) là vượt quá giới hạn phần cứng của Tesla T4 (1024 threads/block).

In [7]:
!./HW1_P1.out ./Earthmap1000x500.pnm earthmap_32_32 32 32

Image size (width x height): 1000 x 500

Processing time (use host): 10.777312 ms

GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use device): 4.863520 ms

Error between device result and host result: 0.000338


In [8]:
!./HW1_P1.out ./ladybug800x533.pnm ladybug_32_32 32 32

Image size (width x height): 800 x 533

Processing time (use host): 5.847072 ms

GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use device): 4.003968 ms

Error between device result and host result: 0.000019


### Giải thích cho sai số nhỏ giữa kết quả của kernel và host function


## Câu 2

In [ ]:
!apt install imagemagick

In [9]:
!nvcc -arch=sm_{major}{minor} HW1_P2.cu -o HW1_P2.out

In [10]:
!./HW1_P2.out ./in.pnm out2_16_16 out2_target.pnm 16 16

Image size (width x height): 512 x 512

argv[1]= ./in.pnm
argv[2]= out2_16_16
argv[3]= out2_target.pnm
Processing time (use host): 216.002563 ms
Error: 0.499257

GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use device): 6.071104 ms
Error: 0.499257

